sdk/python/feast/infra/utils/aws_utils.py

```diff
@@ -0,0 +1,286 @@
+    Raises RedshiftCredentialsError if the statement couldn't be executed due to the validation error.
+    Raises RedshiftQueryError if the query runs but finishes with errors.
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param query: The SQL query to execute
+    :return: Statement ID
+    """
+    statement = execute_redshift_statement_async(
+        redshift_data_client, cluster_id, database, user, query
+    )
+    wait_for_redshift_statement(redshift_data_client, statement)
+    return statement["Id"]
+
+
+def get_redshift_statement_result(redshift_data_client, statement_id: str) -> dict:
+    """ Get the Redshift statement result """
+    return redshift_data_client.get_statement_result(Id=statement_id)
+
+
+def copy_df_to_redshift(
+    redshift_data_client,
+    cluster_id: str,
+    database: str,
+    user: str,
+    s3_resource,
+    s3_path: str,
+    iam_role: str,
+    table_name: str,
+    df: pd.DataFrame,
+) -> None:
+    """
+    Uploads a Pandas DataFrame to Redshift as a new table.
+    The caller is responsible for deleting the table when no longer necessary.
+
+    Here's how the upload process works:
+        1. Pandas DataFrame is converted to PyArrow Table
+        2. PyArrow Table is serialized into a Parquet format on local disk
+        3. The Parquet file is uploaded to S3
+        4. The S3 file is uploaded to Redshift as a new table through COPY command
+        5. The local disk & s3 paths are cleaned up
+
+    :param redshift_data_client: Redshift Data API Service client
+    :param cluster_id: Redshift Cluster Identifier
+    :param database: Redshift Database Name
+    :param user: Redshift username
+    :param s3_resource: S3 Resource object
+    :param s3_path: S3 path where the Parquet file is temporarily uploaded
+    :param iam_role: IAM Role for Redshift to assume during the COPY command.
+                     The role must grant permission to read the S3 location.
+    :param table_name: The name of the new Redshift table where we copy the dataframe
+    :param df: The Pandas DataFrame to upload
+    """
+    bucket, key = get_bucket_and_key(s3_path)
+
+    # Convert Pandas DataFrame into PyArrow table and compile the Redshift table schema
+    table = pa.Table.from_pandas(df)
+    column_names, column_types = [], []
+    for field in table.schema:
+        column_names.append(field.name)
+        column_types.append(pa_to_redshift_value_type(str(field.type)))
+    column_query_list = ", ".join(
+        [
+            f"{column_name} {column_type}"
+            for column_name, column_type in zip(column_names, column_types)
+        ]
+    )
+
+    # Create the temporary table with the desired schema
+    create_query = f"CREATE TABLE {table_name}({column_query_list})"
+    execute_redshift_statement(
```

Comment:
Would it make sense to create this table only after writing the parquet file to s3? Otherwise we're creating empty tables if failures occur directly after.